File: algorithms/CSharp/src/Graph/breadth-first-search.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Algorithms.Graph
{
    public class BreadthFirstSearch
    {
        public class Graph
        {
            private List<List<int>> _edges;
            private bool _isDirected;
            private int _totalNodes;

            public Graph(int totalNodes, bool isDirected)
            {
                _isDirected = isDirected;
                _totalNodes = totalNodes;
                _edges = new List<List<int>>();
                for(int i = 0; i <= totalNodes; i++)
                {
                    _edges.Add(new List<int>());
                }
            }

            public void AddEdge(int source, int destination)
            {
                _edges[source].Add(destination);
                if(!_isDirected)
                {
                    _edges[destination].Add(source);
                }
            }

            public List<int> BFS(int source)
            {
                List<int> path = new List<int>();
                Queue<int> q = new Queue<int>();
                bool[] visited = Enumerable.Repeat(false, _totalNodes + 1).ToArray();
               
                q.Enqueue(source);
                visited[source] = true;

                while(q.Count() > 0)
                {
                    int u = q.Dequeue();
                    path.Add(u);

                    foreach(int v in _edges[u])
                    {
                        if(!visited[v])
                        {
                            q.Enqueue(v);
                            visited[v] = true;
                        }
                    }
                }

                return path;
            }
        }

        public static List<int> BFSDriver(int totalNodes, bool isDirected, int source, List<Tuple<int, int>> edges)
        {
            Graph graph = new Graph(totalNodes, isDirected);
            foreach(var edge in edges)
            {
                graph.AddEdge(edge.Item1, edge.Item2);
            }

            return graph.BFS(source);
        }

        public static void Main()
        {
            List<Tuple<int, int>> edges = new List<Tuple<int, int>>()
            {
                new Tuple<int, int>(1, 2),
                new Tuple<int, int>(1, 3),
                new Tuple<int, int>(2, 3),
                new Tuple<int, int>(3, 1),
                new Tuple<int, int>(3, 4),
                new Tuple<int, int>(4, 4),
            };

            var results = BFSDriver(4, true, 3, edges);
            Console.WriteLine(string.Join("->", results));
        }
    }
}

File: algorithms/JavaScript/src/graph/breadth-first-search.js
// Program to print BFS traversal from a given source vertex s.
// breadthFirstSearch(graph, source) traverses vertices reachable from source.
// for the implementation we need the queue data structure;

// Queue class
class Queue {
  // Array is used to implement a Queue
  constructor() {
    this.items = [];
  }

  // enqueue function
  enqueue(element) {
    // adding element to the queue
    this.items.push(element);
  }
  // dequeue function
  // removing element from the queue
  // returns underflow when called
  // on empty queue
  dequeue() {
    if (this.isEmpty()) {
      return 'Underflow';
    }
    return this.items.shift();
  }
  // isEmpty function
  isEmpty() {
    // return true if the queue is empty.
    return this.items.length == 0;
  }
}


class Graph {
  // defining vertex array and
  // adjacent list
  constructor(noOfVertices) {
    this.noOfVertices = noOfVertices;
    this.AdjList = new Map();
  }
  // add the requeired vertex in the graph
  addVertex(v) {
    this.AdjList.set(v, []);
  }
  // addition of edges in the graph as the added edge are bidirectional
  addEdge(v, w) {
    this.AdjList.get(v).push(w);
    this.AdjList.get(w).push(v);
  }
}
const breadthFirstSearch = (g, source) => {
  const visited = {};
  const q = new Queue();
  visited[source] = true;
  q.enqueue(source);

  while (!q.isEmpty()) {
    // Dequeue a vertex from queue and print it
    const getQueueElement = q.dequeue();
    console.log(getQueueElement);
    const getList = g.AdjList.get(getQueueElement);
    // Get all adjacent vertices of the dequeued
    // vertex s. If a adjacent has not been visited,
    // then mark it visited and enqueue it
    for (let i = 0; i < getList.length; i++) {
      const neigh = getList[i];
      if (!visited[neigh]) {
        visited[neigh] = true;
        q.enqueue(neigh);
      }
    }
  }
};

const g = new Graph(6);
// adding vertices
for (let i = 1; i <= 6; i++) {
  g.addVertex(i);
}
g.addEdge(1, 2);
g.addEdge(2, 4);
g.addEdge(3, 1);
g.addEdge(1, 4);
g.addEdge(5, 3);
g.addEdge(6, 3);

breadthFirstSearch(g, 6);
// when we start bfs for the given graph from point 6 the output is as follow;
// output 6 3 1 5 2 4

// Time complexity: O(n), where n is the number of vertices in graph
// Space complexity: O(n), where n is the number of vertices in graph
